README.md
# EctoAutoFilter

**Automatic Filters based Ecto Schemas**

EctoAutoFilter is a helper for projects that use Ecto Schemas and segregate the queries in entity repository modules. EctoAutoFilter inject the `filter/3` function that by default has a pattern matching for each field of declared entity.

## Installation

EctoAutoFilter can be installed
by adding `ecto_auto_filter` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:ecto_auto_filter, "~> 1.0.0"}
  ]
end
```

## Get Started
After the installation of dependency, we will add the configuration below so that EctoAutoFilter knows what `repo` module it should execute on queries.

```elixir
# config.exs
config :ecto_auto_filter,
  repo: MyApp.Repo,
```

This configuration apply the `MyApp.Repo` for all entity repositories that use EctoAutoFilter. There are also the possibility of injecting a different `repo` setup for a specific entity repository, for example:

```elixir
# config.exs
defmodule UserRepository do
  use EctoAutoFilter,
    repo: MyApp.AnotherRepo,
    schema: MyApp.User
end
```
Now we will declare the entity repository. Consider the `User` schema.

```elixir
defmodule MyApp.User do
  use Ecto.Schema

  schema "users" do
    field(:name, :string)
    field(:email, :string)
    field(:age, :integer)
  end
  ...
end
```
Declare the UserRepository using EctoAutoFilter, passing in `schema` parameter the `User` schema and here we will consider that we configured a global `repo`.

```elixir
defmodule UserRepository do
  use EctoAutoFilter,
    schema: MyApp.User
end
```

Done! We just need to test now.

```elixir
iex> user_one = Repo.insert!(%User{name: "John Doe", email: "john@doe.com", age: 30})
iex> user_two = Repo.insert!(%User{name: "Jane Doe", email: "jane@doe.com", age: 28})
iex> UserRepository.filter(%{age: {30, ">="}})
{:ok, 
  [%User{
    __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
    age: 30,
    email: "john@doe.com",
    id: 1,
    name: "John Doe"
  }]
}
```

## The `filter/3` function
```elixir
@spec filter(map | list, atom, module) :: 
  {:ok, one_result | list} | 
  {:error, :not_found | :filter_not_found | :unsupported_filter } 
def filter(filter, result_format \\ :many, query \\ @base_schema)
```
- `filter`: The first parameter is a map or keywordlist for the filter declaration;
- `result_format`: It's an atom that can be `:one`, `:first` or `:many`, defined by default `:many`;
- `query`: Queriable structure that by default is the base entity.


#### Filter structure

The structure of the filter is a map or a keywordlist in which the keys are field names of the schema to be filtered, e.g: `filter(%{name: "John Doe"})`.
For each key in the filter parameter, EctoAutoFilter concatenate a `where` clause to the `and` logical connector, thus it is possible to compose more complex filters, such as: `filter(%{name: "John Doe", age: 30})`. The filters can receive values directly to an exact value comparison, or values with functions that varing between binary operators or functions such as: `like`, `ilike`, `not_in` or `in`.

- {list, "in | not_in"}: This clause is composed by a tuple with a list and a string `"in"`;
- {value, "> | < | >= | <= | == | !="}: Arithmetical and logical operators;
- {value, "like | ilike"}: Partial search for strings.


### Returning only one value
```elixir
iex> UserRepository.filter(%{age: {28, "=="}}, :one)
{:ok, 
  %User{
    __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
    age: 28,
    email: "jane@doe.com",
    id: 2,
    name: "Jane Doe"
  }
}
```

### Returning the first found value
```elixir
iex> UserRepository.filter(%{email: {"%doe.com", "like"}}, :first)
{:ok,
  %User{
    __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
    age: 30,
    email: "john@doe.com",
    id: 1,
    name: "John Doe"
  }
}
```

### Returning all the found values
```elixir
iex> UserRepository.filter(%{email: {"%doe.com", "like"}})
{:ok,[
  %User{
    __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
    age: 30,
    email: "john@doe.com",
    id: 1,
    name: "John Doe"
  },
  %User{
    __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
    age: 28,
    email: "jane@doe.com",
    id: 2,
    name: "Jane Doe"
  }
]}
```
## Filters

### By value
```elixir
iex> UserRepository.filter(%{age: {28, "=="}})
{:ok,[
  %User{
    __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
    age: 28,
    email: "jane@doe.com",
    id: 2,
    name: "Jane Doe"
  }
]}
```
### Between filters
One key of filter can receive a tuple list in which the EctoAutoFilter concatenate by `and` logical operator, such as the example below where we filter the users from 27 to 29 years old.

```elixir
iex> user_three = Repo.insert!(%User{name: "Tom Doe", email: "tom@doe.com", age: 26})
iex> UserRepository.filter(%{age: [{30, "<"}, {26, ">"}]})
{:ok,[
  %User{
    __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
    age: 28,
    email: "jane@doe.com",
    id: 2,
    name: "Jane Doe"
  }
]}
```

### `in` & `not_in` filter
One key of filter can receive a tuple with a list of values in which the EctoAutoFilter uses `in` or `not_in` to verify the existence or inexistence of the key inside the values from the list.

```elixir
iex> UserRepository.filter(%{id: {[2, 3], "in"}})
{:ok,[
  %User{
    __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
    age: 28,
    email: "jane@doe.com",
    id: 2,
    name: "Jane Doe"
  },
  %User{
    __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
    age: 26,
    email: "tom@doe.com",
    id: 3,
    name: "Tom Doe"
  }
]}

iex> UserRepository.filter(%{id: {[2, 3], "not_in"}})
{:ok,[
  %User{
    __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
    age: 30,
    email: "john@doe.com",
    id: 1,
    name: "John Doe"
  }
]}
```

### `like` & `ilike` filters
```elixir
iex> user_four = Repo.insert!(%User{name: "Norton Doe", email: "norton@doe.com", age: 60})
iex> UserRepository.filter(%{name: {"%to%", "ilike"}})
{:ok,[
  %User{
    __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
    age: 26,
    email: "tom@doe.com",
    id: 3,
    name: "Tom Doe"
  },
  %User{
    __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
    age: 60,
    email: "norton@doe.com",
    id: 4,
    name: "Norton Doe"
  }
]}
```

## Errors
- `{:error, :unsupported_filter}`: It occurs when an unknown key is passed inside a filter, returning this error, even if the other keys are correct.

- `{:error, :filter_not_found}`: It is returned in case the filter is empty. 

- `{:error, :unsupported_filter_operator}`: It is returned if one tuple value is passed with an unknown operator.


## Custom Filters
EctoAutoFilter can be extended through its macro `add_filter/4`, see more in [`add_filter/4`](https://hexdocs.pm/ecto_auto_filter/EctoAutoFilter.html#add_filter/4)


## Contributing

Feedback, feature requests, and fixes are welcomed and encouraged. Please make appropriate use of [Issues](https://github.com/brunolouvem/ecto_auto_filter/issues) and [Pull Requests](https://github.com/brunolouvem/ecto_auto_filter/pulls). All code should have accompanying tests.


CHANGELOG.md
## [0.1.1] - 2021-07-29

- Add naive_datetime field type.
## [0.1.0] - 2021-01-11

- Initial release with base functionality.

test/support/repo.ex
defmodule EctoAutoFilter.Test.Repo do
  @moduledoc false
  use Ecto.Repo,
    otp_app: :ecto_auto_filter,
    adapter: Ecto.Adapters.Postgres
end


test/support/user.ex
defmodule EctoAutoFilter.Test.User do
  @moduledoc false
  use Ecto.Schema

  schema "users" do
    field(:name, :string)
    field(:email, :string)
    field(:age, :integer)
    field(:birth_date, :date)
    field(:occupation, :string)
    field(:nickname, :string)
  end
end


test/test_helper.exs
ExUnit.start()

EctoAutoFilter.Test.Repo.start_link()


mix.exs
defmodule EctoAutoFilter.MixProject do
  use Mix.Project

  @version "0.1.1"
  def project do
    [
      app: :ecto_auto_filter,
      version: @version,
      elixir: "~> 1.8",
      elixirc_paths: elixirc_paths(Mix.env()),
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      aliases: aliases(),
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test
      ],

      # Hex
      description: "A Helpers Set for build automatic filters on Ecto Schemas",
      package: package(),

      # Docs
      name: "Ecto Auto Filter",
      docs: docs()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_env), do: ["lib"]

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:ecto, "~> 3.2"},
      {:ecto_sql, "~> 3.4", only: [:dev, :test]},
      {:excoveralls, "~> 0.10", only: :test},
      {:postgrex, ">= 0.0.0", only: [:dev, :test]},
      {:inch_ex, github: "rrrene/inch_ex", only: [:dev, :test]},
      {:credo, "~> 1.4", only: [:test, :dev], runtime: false},
      {:dialyxir, "~> 1.0", only: [:test, :dev], runtime: false},
      {:ex_doc, "~> 0.22", only: [:dev, :test], runtime: false}
    ]
  end

  defp package do
    [
      maintainers: ["Bruno Louvem"],
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/brunolouvem/ecto_auto_filter"},
      files: ~w(.formatter.exs mix.exs README.md CHANGELOG.md lib/)
    ]
  end

  defp docs do
    [
      main: "readme",
      source_ref: "v#{@version}",
      canonical: "http://hexdocs.pm/ecto_auto_filter",
      source_url: "https://github.com/brunolouvem/readme",
      extras: ["README.md"]
    ]
  end

  defp aliases do
    [
      setup: ["deps.get", "ecto.setup"],
      "ecto.setup": ["ecto.create", "ecto.migrate", "run test/support/seeds.exs"],
      "ecto.reset": ["ecto.drop", "ecto.setup"],
      ci: [
        "format --check-formatted",
        "credo --strict",
        "test --raise",
        "dialyzer"
      ],
      "test.ci": [
        "ecto.setup",
        "ci"
      ]
    ]
  end
end


lib/ecto_auto_filter.ex
defmodule EctoAutoFilter do
  @moduledoc """
  Ecto Auto Filter
  **Automatic Filters based Ecto Schemas**

  EctoAutoFilter is a helper for projects that use Ecto Schemas and segregate the queries in entity repository modules.
  EctoAutoFilter inject the `filter/3` function that by default has a pattern matching for each field of declared entity.
  """
  @all_binaries_coparable_types ~w(id date integer datetime naive_datetime)a

  @doc """
  Injeta a função `filter/3` e as funções privadas responsáveis pelo filtro customizado a partir do `schema` base passado

  Injects the `filter/3` functions and the privated functions responsible for custom filters from base entity schema's passed.

  Options:
   - `schema`: This options is required and is necessary for EctoAutoFilter build the queries.
   - `repo`: This options is optional and is used to execute the queries, and when is passed such as `use` option, overrides the global config for the current module.

  Exemple:
      defmodule MyApp.User do
        use Ecto.Schema

        schema "users" do
          field(:name, :string)
          field(:email, :string)
          field(:age, :integer)
        end
        ...
      end


      defmodule UserRepository do
        use EctoAutoFilter,
          repo: MyApp.Repo,
          schema: MyApp.User
      end

  After that entity repository module is declared the filter functions can already be used like a below example:

  Depois de declarado o using no módulo de repositório as funções de filtro podem ser usadas como no exemplo a seguir:
      iex> user_one = Repo.insert!(%User{name: "John Doe", email: "john@doe.com", age: 30})
      iex> user_two = Repo.insert!(%User{name: "Jane Doe", email: "jane@doe.com", age: 28})
      iex> UserRepository.filter(%{age: {30, ">="}})
      %User{
        __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
        age: 30,
        email: "john@doe.com",
        id: 1,
        name: "John Doe"
      }
  """
  defmacro __using__(opts) do
    quote bind_quoted: [opts: opts] do
      @before_compile EctoAutoFilter

      import Ecto.Query
      import EctoAutoFilter

      @repo opts[:repo] || Application.compile_env(:ecto_auto_filter, :repo)
      @schema opts[:schema]

      @doc false
      def filter(filter, result_format \\ :many, query \\ @schema)

      def filter(filter, _result_format, _query) when filter == %{} or filter == [],
        do: {:error, :filter_not_found}

      def filter(filters, result_format, query) when is_map(filters) or is_list(filters) do
        cond do
          is_map(filters) ->
            build_filters(filters, result_format, query)

          is_list(filters) and Keyword.keyword?(filters) ->
            build_filters(filters, result_format, query)

          true ->
            {:error, :unsupported_filter}
        end
      end

      def filter(_filter, _result_format, _query), do: {:error, :unsupported_filter}

      defp build_filters(filters, result_format, query) do
        Enum.reduce(filters, query, fn
          {field, value}, query ->
            apply_filter(field, query, value)
        end)
        |> run_query(result_format)
        |> handle_result()
      end

      for schema_field <- @schema.__schema__(:fields) do
        field_type = @schema.__schema__(:type, schema_field)

        write_filters_by_type(schema_field, field_type)
        |> Enum.each(fn macro ->
          macro
          |> Macro.expand(__ENV__)
          |> Code.eval_quoted()
        end)
      end

      defp run_query({:error, _} = error_tuple, _result_format), do: error_tuple

      defp run_query(queryable, :one) do
        @repo.one(queryable)
      end

      defp run_query(queryable, :first) do
        @repo.all(queryable)
        |> case do
          [first | _] ->
            first

          _ ->
            nil
        end
      end

      defp run_query(queryable, :many) do
        @repo.all(queryable)
      end

      defp handle_result({:error, _} = error_tuple), do: error_tuple
      defp handle_result(nil), do: {:error, :not_found}
      defp handle_result(result), do: {:ok, result}
    end
  end

  @doc false
  def write_filters_by_type(field_name, field_type)
      when field_type in @all_binaries_coparable_types do
    [
      list_filters(field_name),
      not_in_compare_filter(field_name),
      in_compare_filter(field_name),
      binary_compare_filter(field_name),
      equal_compare_filter(field_name)
    ]
  end

  @doc false
  def write_filters_by_type(field_name, _) do
    [
      list_filters(field_name),
      not_in_compare_filter(field_name),
      in_compare_filter(field_name),
      ilike_filter(field_name),
      like_filter(field_name),
      equal_compare_filter(field_name)
    ]
  end

  defp list_filters(field_name) do
    quote do
      defp apply_filter(unquote(field_name) = field_id, query, values) when is_list(values) do
        Enum.reduce(values, query, fn value, query_acc ->
          apply_filter(field_id, query_acc, value)
        end)
      end
    end
  end

  defp equal_compare_filter(field_name) do
    quote do
      defp apply_filter(unquote(field_name), query, value) do
        where(query, [r], field(r, ^unquote(field_name)) == ^value)
      end
    end
  end

  defp not_in_compare_filter(field_name) do
    quote do
      defp apply_filter(unquote(field_name), query, {values, "not_in"}) when is_list(values) do
        where(query, [r], field(r, ^unquote(field_name)) not in ^values)
      end
    end
  end

  defp in_compare_filter(field_name) do
    quote do
      defp apply_filter(unquote(field_name), query, {values, "in"}) when is_list(values) do
        where(query, [r], field(r, ^unquote(field_name)) in ^values)
      end
    end
  end

  defp ilike_filter(field_name) do
    quote do
      defp apply_filter(unquote(field_name), query, {value, "ilike"}) when is_binary(value) do
        where(query, [r], ilike(field(r, ^unquote(field_name)), ^value))
      end
    end
  end

  defp like_filter(field_name) do
    quote do
      defp apply_filter(unquote(field_name), query, {value, "like"}) when is_binary(value) do
        where(query, [r], like(field(r, ^unquote(field_name)), ^value))
      end
    end
  end

  defp binary_compare_filter(field_name) do
    quote do
      defp apply_filter(unquote(field_name), query, {value, op}) when is_binary(op) do
        case op do
          "==" -> where(query, [r], field(r, ^unquote(field_name)) == ^value)
          "!=" -> where(query, [r], field(r, ^unquote(field_name)) != ^value)
          ">=" -> where(query, [r], field(r, ^unquote(field_name)) >= ^value)
          "<=" -> where(query, [r], field(r, ^unquote(field_name)) <= ^value)
          "<" -> where(query, [r], field(r, ^unquote(field_name)) < ^value)
          ">" -> where(query, [r], field(r, ^unquote(field_name)) > ^value)
          _ -> {:error, :unsupported_filter_operator}
        end
      end
    end
  end

  @doc """
  Adiciona um filtro customizado e componivel aos demais filtros do módulo base em tempo de compilação.
  - query: é queryable que a função irá receber;
  - value: é o valor que será recebido para compor o filtro e pode ser manipulado por pattern matching;
  - key: indentify the created filter in `filter/3` function.

  Este filtro é adicionado as demais regras de filtro e se torna disponível na forma de `filter/3`,
  por exemplo:
      defmodule MyApp.User do
        use Ecto.Schema

        schema "users" do
          field(:name, :string)
          field(:email, :string)
          field(:age, :integer)
          field(:bith_date, :date)
        end
        ...
      end


      defmodule UserRepository do
        use EctoAutoFilter,
          repo: MyApp.Repo,
          schema: MyApp.User

        add_filter query, value, :birth_years_ago do
          x_years_ago = (365 * value)
          limit_date = Date.utc_today() |> Date.add(-x_years_ago)

          where(query, [r], r.birth_date == ^limit_date)
        end
      end

  Depois de declarado o filtro customizado ele pode ser usado conforme o exemplo a seguir:
      iex> user_one = Repo.insert!(%User{name: "John Doe", email: "john@doe.com", age: 30, birth_date: 1991-01-01})
      iex> user_two = Repo.insert!(%User{name: "Jane Doe", email: "jane@doe.com", age: 28, birth_date: 1993-01-01})
      iex> UserRepository.filter(%{birth_years_ago: 28})
      %User{
        __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
        age: 28,
        email: "jane@doe.com",
        id: 2,
        name: "Jane Doe"
      }
  """
  defmacro add_filter(query, value, key, do: block) do
    quote do
      defp apply_filter(unquote(key), unquote(query), unquote(value)) do
        unquote(block)
      end
    end
  end

  defmacro __before_compile__(_env) do
    quote do
      defp apply_filter(_, _, _), do: {:error, :unsupported_filter}
    end
  end
end
